<a href="https://colab.research.google.com/github/cdtalley/Data-Science-Portfolio/blob/main/Jane_Street_Market_Prediction_XGBoost_and_Hyperparameter_Tuning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Jane Street Market Prediction: XGBoost and Hyperparameter Tuning

## Using Financial Market Data to Predict Successful Trades

Financial market data is a driving force behind the decisions of traders everyday. As data collection and processing capabilites grow, and the amount of electronic trading increases with the usage of free trading apps and mechanical/algorithmic based trading, market data and trading data captured from the markets grows in complexity each day. This allows for development of machine learning models that can predict with some degree of accuracy the decision to make on a trade, to pass or initiate the buy of an asset that will give us a positive return on our trade at a later date.

### About Our Data

This dataset contains an anonymized set of features, feature_{0...129}, representing real stock market data. Each row in the dataset represents a trading opportunity, for which I will be predicting an action value: 1 to make the trade and 0 to pass on it. Each trade has an associated weight and resp, which together represents a return on the trade. The date column is an integer which represents the day of the trade, while ts_id represents a time ordering. In addition to anonymized feature values, you are provided with metadata about the features in features.csv.

In the training set, train.csv, we are provided a resp value, as well as several other resp_{1,2,3,4} values that represent returns over different time horizons. These variables are not included in the test set. Trades with weight = 0 were intentionally included in the dataset for completeness.

* train.csv - the training set, contains historical data and returns
* features.csv - metadata pertaining to the anonymized features


Data is from the Jane Street Market Prediction Dataset, linked here; https://www.kaggle.com/c/jane-street-market-prediction

# Exploratory Data Analysis

In [1]:
# Importing necessary data visualization and machine learning packages and .csv file into a pandas DataFrame.
%matplotlib inline
import time
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
from scipy import stats
from google.colab import drive
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import GridSearchCV
import plotly as py
import plotly.graph_objs as go
import plotly.tools as tls
from plotly.offline import iplot, init_notebook_mode
import cufflinks
import cufflinks as cf
import plotly.figure_factory as ff
import os
import warnings

# Suppress warnings.
warnings.filterwarnings("ignore")

In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=False)

Mounted at /content/drive


In [2]:
# Importing pandas DataFrame.
train = pd.read_csv("/content/drive/My Drive/Data/train.csv")

In [3]:
gpu_info = !nvidia-smi
gpu_info = '\n'.join(gpu_info)
if gpu_info.find('failed') >= 0:
  print('Select the Runtime > "Change runtime type" menu to enable a GPU accelerator, ')
  print('and then re-execute this cell.')
else:
  print(gpu_info)

Mon Feb  8 05:38:08 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.39       Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   36C    P0    26W / 300W |      0MiB / 16130MiB |      0%      Default |
|                               |                      |                 ERR! |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
# Viewing DataFrame column types to make sure data types are correct.
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2390491 entries, 0 to 2390490
Columns: 138 entries, date to ts_id
dtypes: float64(135), int64(3)
memory usage: 2.5 GB


In [ ]:
# Calling pandas head function to take a glance at our data.
train.head()

date     weight    resp_1  ...  feature_128  feature_129  ts_id
0     0   0.000000  0.009916  ...     2.301488    11.445807      0
1     0  16.673515 -0.002828  ...    -1.304614     1.898684      1
2     0   0.000000  0.025134  ...     6.638248     9.427299      2
3     0   0.000000 -0.004730  ...     3.856384     1.013469      3
4     0   0.138531  0.001252  ...     0.362636     3.926633      4

[5 rows x 138 columns]

In [ ]:
# Setting max rows to show all percent missing values.
pd.set_option("max_rows", None)

In [ ]:
# Finding the percent of missing values by calling pandas .isna function which returns a mask of bool values for each element in 
# DataFrame that indicates whether an element is not an NA value, and rounding the mean.
percent_missing = train.isna().mean().round(4) * 100
percent_missing = pd.DataFrame(percent_missing)
percent_missing

0
date          0.00
weight        0.00
resp_1        0.00
resp_2        0.00
resp_3        0.00
resp_4        0.00
resp          0.00
feature_0     0.00
feature_1     0.00
feature_2     0.00
feature_3     0.02
feature_4     0.02
feature_5     0.00
feature_6     0.00
feature_7    16.45
feature_8    16.45
feature_9     0.03
feature_10    0.03
feature_11    3.35
feature_12    3.35
feature_13    0.64
feature_14    0.64
feature_15    0.28
feature_16    0.28
feature_17   16.55
feature_18   16.55
feature_19    0.03
feature_20    0.03
feature_21    3.41
feature_22    3.41
feature_23    0.64
feature_24    0.64
feature_25    0.28
feature_26    0.28
feature_27   16.55
feature_28   16.55
feature_29    0.03
feature_30    0.03
feature_31    3.41
feature_32    3.41
feature_33    0.64
feature_34    0.64
feature_35    0.28
feature_36    0.28
feature_37    0.00
feature_38    0.00
feature_39    0.00
feature_40    0.00
feature_41    0.00
feature_42    0.00
feature_43    0.00
feature_44    0.02
feature_45    0.02
feature_46    0.00
feature_47    0.00
feature_48    0.00
feature_49    0.00
feature_50    0.00
feature_51    0.00
feature_52    0.00
feature_53    0.00
feature_54    0.00
feature_55    2.86
feature_56    0.03
feature_57    0.00
feature_58    0.00
feature_59    0.00
feature_60    0.00
feature_61    0.00
feature_62    0.00
feature_63    0.00
feature_64    0.00
feature_65    0.00
feature_66    0.00
feature_67    0.00
feature_68    0.00
feature_69    0.00
feature_70    0.00
feature_71    0.00
feature_72   14.70
feature_73    0.03
feature_74    2.68
feature_75    0.64
feature_76    0.28
feature_77    0.00
feature_78   14.70
feature_79    0.03
feature_80    2.68
feature_81    0.64
feature_82    0.28
feature_83    0.00
feature_84   14.70
feature_85    0.03
feature_86    2.68
feature_87    0.64
feature_88    0.28
feature_89    0.00
feature_90   14.70
feature_91    0.03
feature_92    2.68
feature_93    0.64
feature_94    0.28
feature_95    0.00
feature_96   14.70
feature_97    0.03
feature_98    2.68
feature_99    0.64
feature_100   0.28
feature_101   0.00
feature_102  14.70
feature_103   0.03
feature_104   2.68
feature_105   0.64
feature_106   0.28
feature_107   0.00
feature_108  14.70
feature_109   0.03
feature_110   2.68
feature_111   0.64
feature_112   0.28
feature_113   0.00
feature_114  14.70
feature_115   0.03
feature_116   2.68
feature_117   0.64
feature_118   0.28
feature_119   0.00
feature_120   2.92
feature_121   2.92
feature_122   0.01
feature_123   0.01
feature_124   0.67
feature_125   0.67
feature_126   0.37
feature_127   0.37
feature_128   0.08
feature_129   0.08
ts_id         0.00

Some of our features are missing a good amount of data, but this is still an acceptable amount. We can proceed further with machine learning analysis.

In [4]:
# Importing pandas DataFrame.
features =  pd.read_csv("/content/drive/My Drive/Data/features.csv")

In [ ]:
# Viewing DataFrame column types to make sure data types are correct.
features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 30 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   feature  130 non-null    object
 1   tag_0    130 non-null    bool  
 2   tag_1    130 non-null    bool  
 3   tag_2    130 non-null    bool  
 4   tag_3    130 non-null    bool  
 5   tag_4    130 non-null    bool  
 6   tag_5    130 non-null    bool  
 7   tag_6    130 non-null    bool  
 8   tag_7    130 non-null    bool  
 9   tag_8    130 non-null    bool  
 10  tag_9    130 non-null    bool  
 11  tag_10   130 non-null    bool  
 12  tag_11   130 non-null    bool  
 13  tag_12   130 non-null    bool  
 14  tag_13   130 non-null    bool  
 15  tag_14   130 non-null    bool  
 16  tag_15   130 non-null    bool  
 17  tag_16   130 non-null    bool  
 18  tag_17   130 non-null    bool  
 19  tag_18   130 non-null    bool  
 20  tag_19   130 non-null    bool  
 21  tag_20   130 non-null    bool  
 22  ta

In [ ]:
# Calling pandas head function to take a glance at our data.
features.head()

feature  tag_0  tag_1  tag_2  ...  tag_25  tag_26  tag_27  tag_28
0  feature_0  False  False  False  ...   False   False   False   False
1  feature_1  False  False  False  ...   False   False   False   False
2  feature_2  False  False  False  ...   False   False   False   False
3  feature_3  False  False  False  ...   False   False   False   False
4  feature_4  False  False  False  ...   False   False   False   False

[5 rows x 30 columns]

# Target Variable and Train-Test-Split

In [9]:
# Disregarding trades that do not contribute to scoring with a weight of 0 for training set.
train = train[train['weight'] != 0]
# Creating query to train data on certain date range of 85 days.
train = train.query('date > 85').reset_index(drop = True) 
# Limiting memory usage by converting float 64 to float 32 in applicable data types.
train = train.astype({c: np.float32 for c in train.select_dtypes(include='float64').columns}) #limit memory use

# Creating target variable by multiplying weight and resp values and selecting those with a positive return.
train['action'] = ((train['weight'].values * train['resp'].values) > 0).astype('int')

# Filling missing values with the mean.
train.fillna(train.mean(),inplace=True)

cols = [c for c in train.columns if 'feature' in c]

# Creating independent variables by extracting feature columns to create X variable.
X = train.loc[:, train.columns.str.contains('feature')]
# Creating our target variable using our newly created action variable.
Y = train.loc[:, 'action']

In [ ]:
# Creating a simple plot to show class balance.
x = train['action'].value_counts().index
y = train['action'].value_counts().values

trace2 = go.Bar(
     x=x ,
     y=y,
     marker=dict(
         color=y,
         colorscale = 'Viridis',
         reversescale = True
     ),
     name="Imbalance",    
 )
layout = dict(
     title="Class Balance for Target Variable",
     #width = 900, height = 500,
     xaxis=go.layout.XAxis(
     automargin=True),
     yaxis=dict(
         showgrid=False,
         showline=False,
         showticklabels=True,
 #         domain=[0, 0.85],
     ), 
)
fig1 = go.Figure(data=[trace2], layout=layout)
iplot(fig1)

The target variable is balanced and does not need any further processing to fix class imbalance.

In [10]:
# Using sklearns train_test_split to split data into random train and test subsets to calculate accuracy.
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.20, random_state=42)

# XGBoost Classifier

First we will set up a simple XGBoost classification model without hyperparameter tuning to see how performance is increased with RandomSearchCV.

In [11]:
import xgboost as xgb
print("XGBoost version:", xgb.__version__)

XGBoost version: 0.90


In [ ]:
clf = xgb.XGBClassifier(
    random_state=2021,
    tree_method='gpu_hist'  # Needed for GPU usage.
)

In [ ]:
%time clf.fit(X_train, y_train)

CPU times: user 3.49 s, sys: 1.53 s, total: 5.02 s
Wall time: 5.41 s


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=2021,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, tree_method='gpu_hist', verbosity=1)

In [ ]:
# Print classification report for our XGBoost classifier.
y_pred = clf.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)
report = classification_report(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

print("Accuracy: ", accuracy)
print("Classification report:")
print(report)
print("Confusion matrix:")
print(cm)

Accuracy:  0.5278141038490787
Classification report:
              precision    recall  f1-score   support

           0       0.53      0.48      0.50    156239
           1       0.53      0.57      0.55    158044

    accuracy                           0.53    314283
   macro avg       0.53      0.53      0.53    314283
weighted avg       0.53      0.53      0.53    314283

Confusion matrix:
[[75013 81226]
 [67174 90870]]


In [ ]:
# Print cross validation score.
cross_val_score(clf, X_train, y_train, cv=5)

array([0.5290124 , 0.52668568, 0.52656448, 0.52629402, 0.52789688])

Model performance metrics leave much more to be desired. We will adjust hyperparameters with further models to demonstrate the importance of hyperparameter tuning.

# XGBoost: Hyperparameter Tuning Using RandomizedSearchCV

In [ ]:
from sklearn.model_selection import RandomizedSearchCV

# Create parameter grid for RandomizedSearchCV.
parameters = {'n_estimators':[1000,2000],
              'max_depth':[12,22],
              'learning_rate':[0.03,0.06],
              'subsample':[0.5,1],
              'colsample_bytree':[1,0.5],
              'gamma':[0.5,1],
              'min_child_weight':[1,2],
              'tree_method':['gpu_hist']}

# Create decision tree classifier with RandomizedSearchCV.
CV_clf = RandomizedSearchCV(clf, parameters, random_state=0)

# Fitting decision tree classifier to training data.
CV_clf.fit(X_train, y_train)

RandomizedSearchCV(cv=None, error_score=nan,
                   estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                           colsample_bylevel=1,
                                           colsample_bynode=1,
                                           colsample_bytree=1, gamma=0,
                                           learning_rate=0.1, max_delta_step=0,
                                           max_depth=3, min_child_weight=1,
                                           missing=None, n_estimators=100,
                                           n_jobs=1, nthread=None,
                                           objective='binary:logistic',
                                           random_state=2021, reg_alpha=0,
                                           reg_lambd...
                                           verbosity=1),
                   iid='deprecated', n_iter=10, n_jobs=None,
                   param_distributions={'colsample_bytree': 

In [ ]:
# Calling .best_params_ to pull best parameters from RandomizedSearchCV.
CV_clf.best_params_

{'colsample_bytree': 0.5,
 'gamma': 1,
 'learning_rate': 0.06,
 'max_depth': 12,
 'min_child_weight': 2,
 'n_estimators': 1000,
 'subsample': 1,
 'tree_method': 'gpu_hist'}

In [ ]:
clf = xgb.XGBClassifier(
    random_state=0,
    tree_method='gpu_hist',  # Needed for GPU usage.
    colsample_bytree=0.5,
    gamma=1,
    learning_rate=0.06,
    max_depth=12,
    min_child_weight=2,
    n_estimators=1000,
    subsample=1,
)

In [ ]:
%time clf.fit(X_train, y_train)

CPU times: user 1min 19s, sys: 13 s, total: 1min 32s
Wall time: 1min 32s


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5, gamma=1,
              learning_rate=0.06, max_delta_step=0, max_depth=12,
              min_child_weight=2, missing=None, n_estimators=1000, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, tree_method='gpu_hist', verbosity=1)

In [ ]:
# Print classification report for our XGBoost classifier.
y_pred = clf.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)
report = classification_report(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

print("Accuracy: ", accuracy)
print("Classification report:")
print(report)
print("Confusion matrix:")
print(cm)

Accuracy:  0.6407378063719641
Classification report:
              precision    recall  f1-score   support

           0       0.64      0.63      0.64    156239
           1       0.64      0.65      0.64    158044

    accuracy                           0.64    314283
   macro avg       0.64      0.64      0.64    314283
weighted avg       0.64      0.64      0.64    314283

Confusion matrix:
[[ 98847  57392]
 [ 55518 102526]]


In [ ]:
# Print cross validation score.
cross_val_score(clf, X_train, y_train, cv=5)

array([0.62702892, 0.62617778, 0.6226325 , 0.6273138 , 0.62595356])

We were able to increase model performance by 10%, a major increase. This makes our model certainly more viable, but we need to adjust it further with another hyperparameter grid exploring larger numbers of estimators and max_depth.

In [ ]:
# Creating another parameter grid for RandomizedSearchCV.
parameters = {'n_estimators':[1000,2000],
              'max_depth':[12,24],
              'learning_rate':[0.06,0.2],
              'subsample':[1],
              'colsample_bytree':[0.5],
              'gamma':[0.5,1],
              'min_child_weight':[2,4],
              'tree_method':['gpu_hist']}

# Create decision tree classifier with RandomizedSearchCV.
CV_clf = RandomizedSearchCV(clf, parameters, random_state=0)

# Fitting decision tree classifier to training data.
CV_clf.fit(X_train, y_train)

RandomizedSearchCV(cv=None, error_score=nan,
                   estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                           colsample_bylevel=1,
                                           colsample_bynode=1,
                                           colsample_bytree=0.5, gamma=1,
                                           learning_rate=0.06, max_delta_step=0,
                                           max_depth=12, min_child_weight=2,
                                           missing=None, n_estimators=1000,
                                           n_jobs=1, nthread=None,
                                           objective='binary:logistic',
                                           random_state=0, reg_alpha=0,
                                           reg_lam...
                                           verbosity=1),
                   iid='deprecated', n_iter=10, n_jobs=None,
                   param_distributions={'colsample_bytree': 

In [35]:
# Calling .best_params_ to pull best parameters from RandomizedSearchCV.
CV_clf.best_params_

{'colsample_bytree': 0.5,
 'gamma': 1,
 'learning_rate': 0.06,
 'max_depth': 24,
 'min_child_weight': 2,
 'n_estimators': 1000,
 'subsample': 1,
 'tree_method': 'gpu_hist'}

In [36]:
clf = xgb.XGBClassifier(
    random_state=0,
    tree_method='gpu_hist',  # Needed for GPU usage.
    colsample_bytree=0.5,
    gamma=1,
    learning_rate=0.06,
    max_depth=24,
    min_child_weight=2,
    n_estimators=1000,
    subsample=1,
)

In [37]:
%time clf.fit(X_train, y_train)

CPU times: user 13min 2s, sys: 1min, total: 14min 3s
Wall time: 14min 2s


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5, gamma=1,
              learning_rate=0.06, max_delta_step=0, max_depth=24,
              min_child_weight=2, missing=None, n_estimators=1000, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, tree_method='gpu_hist', verbosity=1)

In [38]:
# Print classification report for our XGBoost classifier.
y_pred = clf.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)
report = classification_report(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

print("Accuracy: ", accuracy)
print("Classification report:")
print(report)
print("Confusion matrix:")
print(cm)

Accuracy:  0.6748949195470324
Classification report:
              precision    recall  f1-score   support

           0       0.68      0.67      0.67    156239
           1       0.67      0.68      0.68    158044

    accuracy                           0.67    314283
   macro avg       0.67      0.67      0.67    314283
weighted avg       0.67      0.67      0.67    314283

Confusion matrix:
[[103972  52267]
 [ 49908 108136]]


We can see that increasing our XGBoost models max_depth increases performance slightly again with a model performance increase in accuracy scoring of about 5%. Further testing with increased max_depth up to 100 did not show any signs of model improvement, only longer run times.

# Conclusion

This notebook demonstrates the importance of hyperparameter tuning when it comes to model performance. Our initial model was not showing any useful predictive accuracy, with a scoring of 52%. After hyperparameter tuning we were able to achieve a predictive accuracy of almost 68%, a roughly 16% increase in model accuracy scoring. XGBoost continues to be a useful model for working with large amounts of complex numerical data like the market data here.